### PR TITLE
fix(auth): Use 'owner' token when communicating with Auth emulator

### LINKED ISSUE #1077: Error fetching access token. Not connecting to Firebase Auth Emulator if application_default_credentials.json not present
### [READ] Step 1: Are you in the right place?

  * For issues related to __the code in this repository__ file a Github issue.
  * If the issue pertains to Cloud Firestore, read the instructions in the "Firestore issue"
    template.
  * For general technical questions, post a question on [StackOverflow](http://stackoverflow.com/)
    with the firebase tag.
  * For general Firebase discussion, use the [firebase-talk](https://groups.google.com/forum/#!forum/firebase-talk)
    google group.
  * For help troubleshooting your application that does not fall under one
    of the above categories, reach out to the personalized
    [Firebase support channel](https://firebase.google.com/support/).

### [REQUIRED] Step 2: Describe your environment

  * Operating System version: Debian 10 (but I think is OS independent). Runing in the container produced by this Dockerfile
 ``` Dockerfile
ARG VARIANT="14-buster"
FROM mcr.microsoft.com/vscode/devcontainers/typescript-node:0-${VARIANT}
RUN apt-get update && export DEBIAN_FRONTEND=noninteractive \
    && apt-get -y install --no-install-recommends openjdk-11-jre-headless
RUN echo "deb [signed-by=/usr/share/keyrings/cloud.google.gpg] http://packages.cloud.google.com/apt cloud-sdk main" | tee -a /etc/apt/sources.list.d/google-cloud-sdk.list && curl https://packages.cloud.google.com/apt/doc/apt-key.gpg | apt-key --keyring /usr/share/keyrings/cloud.google.gpg  add - && apt-get update -y && apt-get install google-cloud-sdk -y
RUN sudo -u node npm install -g @nestjs/cli@7.4.0 firebase-tools@v8.14.0 nx@10.3.0
 ```
  * Firebase SDK version: v8.14.0
  * Firebase Product: auth
  * Node.js version: v10.22.0
  * NPM version: 6.14.6

### [REQUIRED] Step 3: Describe the problem

If the 'application_default_credentials.json' is not present in the gcloud config path I get the error from below when I try to call any auth method even though the emulator is started, the FIREBASE_AUTH_EMULATOR_HOST env var is correctly set and the initializeApp is correctly called.

Based on my research this problem comes from the getApplicationDefault function inside the src/credential/credential-internal.ts file. I solved this by running `gcloud auth application-default login` (found on internet, don't know exactly what it does) which generated a valid application_default_credentials.json file

I don't completely understand the underlying mechanisms but I think this behaviour is not great for my use case which is probably very similar with a CI/CD pipeline use case. Also, there's nothing mentioned in the Firebase Auth emulator about this and I really like to be able to run all emulators without a real project but, doing this gcloud config makes me worried not to interact by mistake with the real gcloud project.

```
 Error: Credential implementation provided to initializeApp() via the "credential" property failed to fetch a valid Google OAuth2 access token with the following error: "Error fetching access token: Error while making request: getaddrinfo ENOTFOUND metadata.google.internal metadata.google.internal:80. Error code: ENOTFOUND".
    at FirebaseAppError.FirebaseError [as constructor] (C:\Users\Marian\Desktop\test\tp\node_modules\firebase-admin\lib\utils\error.js:43:28)
    at FirebaseAppError.PrefixedFirebaseError [as constructor] (C:\Users\Marian\Desktop\test\tp\node_modules\firebase-admin\lib\utils\error.js:89:28)
    at new FirebaseAppError (C:\Users\Marian\Desktop\test\tp\node_modules\firebase-admin\lib\utils\error.js:124:28)
    at C:\Users\Marian\Desktop\test\tp\node_modules\firebase-admin\lib\firebase-app.js:124:23
    at process._tickCallback (internal/process/next_tick.js:68:7)
```

#### Steps to reproduce:

1. Run the firebase emulators (especially the auth emulator)
2. Set the FIREBASE_AUTH_EMULATOR_HOST env var as described in the docs
3. Create a simple npm project that uses the firebase-admin
4. Write a simple script (see detailed code below) that makes any call on the auth using the firebase-admin

#### Relevant Code:

I'm running a dev environment using the VSCode .devcontainer feature using the following files:

main.Dockerfile
``` Dockerfile
ARG VARIANT="14-buster"
FROM mcr.microsoft.com/vscode/devcontainers/typescript-node:0-${VARIANT}
RUN apt-get update && export DEBIAN_FRONTEND=noninteractive \
    && apt-get -y install --no-install-recommends openjdk-11-jre-headless
RUN echo "deb [signed-by=/usr/share/keyrings/cloud.google.gpg] http://packages.cloud.google.com/apt cloud-sdk main" | tee -a /etc/apt/sources.list.d/google-cloud-sdk.list && curl https://packages.cloud.google.com/apt/doc/apt-key.gpg | apt-key --keyring /usr/share/keyrings/cloud.google.gpg  add - && apt-get update -y && apt-get install google-cloud-sdk -y
RUN sudo -u node npm install -g @nestjs/cli@7.4.0 firebase-tools@v8.14.0 nx@10.3.0
```

firebase_emulator.Dockerfile
``` Dockerfile
ARG VARIANT="14-buster"
FROM mcr.microsoft.com/vscode/devcontainers/typescript-node:0-${VARIANT}

RUN apt-get update && export DEBIAN_FRONTEND=noninteractive \
    && apt-get -y install --no-install-recommends openjdk-11-jre-headless

RUN sudo -u node npm install -g firebase-tools@v8.14.0

COPY ./firebase.json .
COPY ./.firebaserc .

RUN firebase setup:emulators:database
RUN firebase setup:emulators:firestore

ENTRYPOINT [ "firebase",  "emulators:start"]
```

docker-compose.yml
``` yaml
version: '3'
services:
  main:
    build: 
      context: .
      dockerfile: main.Dockerfile
      args:
        VARIANT: 10
    volumes:
      - ..:/workspace
    command: sleep infinity
    environment:
      FIRESTORE_EMULATOR_HOST: "firebase_emulator:8080"
      FIREBASE_DATABASE_EMULATOR_HOST: "firebase_emulator:9000"
      FIREBASE_AUTH_EMULATOR_HOST: "firebase_emulator:9099"
      GCLOUD_PROJECT: "local"
    links:
      - firebase_emulator
  firebase_emulator:
    build: 
      context: .
      dockerfile: firebase_emulator.Dockerfile
      args:
        VARIANT: 10
    ports:
      - "4000:4000"
      - "5001:5001"
      - "8080:8080"
      - "9000:9000"
      - "8085:8085"
      - "4500:4500"
      - "4400:4400"
      - "9005:9005"
      - "9099:9099"
```

devcontainer.json
``` json
{
	"name": "Node.js & TypeScript & GCP & Firebase & Nest",
	"dockerComposeFile":"docker-compose.yml",
	"service": "main",
	"workspaceFolder": "/workspace",
	"settings": { 
		"terminal.integrated.shell.linux": "/bin/zsh",
		"files.autoSave": "afterDelay",
		"git.autofetch": true,
		"workbench.iconTheme": "material-icon-theme"
	},
	"extensions": [
		"dbaeumer.vscode-eslint",
		"ms-vscode.vscode-typescript-tslint-plugin",
		"eg2.vscode-npm-script",
		"redhat.vscode-yaml",
		"abhijoybasak.nestjs-files",
		"nrwl.angular-console",
		"esbenp.prettier-vscode",
		"firsttris.vscode-jest-runner",
		"github.vscode-pull-request-github",
		"pkief.material-icon-theme",
		"humao.rest-client"
	],
	"remoteEnv": {
		"GITHUB_TOKEN": "${localEnv:GITHUB_TOKEN}",
		"FIREBASE_TOKEN": "${localEnv:FIREBASE_TOKEN}"
	},
	"postCreateCommand": "npm install && npm install --only=dev",
	"remoteUser": "node"
}
```

firebase.json
``` json
{
    "emulators": {
      "firestore": {
        "port": 8080,
        "host": "0.0.0.0"
      },
      "database": {
        "port": 9000,
        "host": "0.0.0.0"
      },
      "auth": {
        "port": 9099,
        "host": "0.0.0.0"
      },
      "ui": {
        "enabled": true,
        "port": 4000,
        "host": "0.0.0.0"
      }
    }
 }
```

.firebaserc
``` json
{
  "projects": {
    "default": "local"
  }
}
```

The minimal code required is to call some auth function using the firebase-admin package. For example:

``` js
// The test script make sure the FIREBASE_AUTH_EMULATOR_HOST environment var is set
const process=require("process")
process.env.FIREBASE_AUTH_EMULATOR_HOST = "localhost:9099";

var admin = require('firebase-admin');

admin.initializeApp({
    projectId: 'local'
})

admin.auth().createUser({});
```

### PR DESCRIPTION
**Pull request by samtstern**
Hey there! So you want to contribute to a Firebase SDK? 
Before you file this pull request, please read these guidelines:

### Discussion

Fixes #1077 

### Testing

Wrote a simple `test.js` script:

```js
const admin = require('firebase-admin');
admin.initializeApp({
  projectId: 'fir-dumpster'
});

async function main() {
  console.log(`FIREBASE_AUTH_EMULATOR_HOST=${process.env.FIREBASE_AUTH_EMULATOR_HOST}`);
  const user = await admin.auth().createUser({
    uid: "user" + new Date().getTime()
  });

  console.log("Created: ", user.uid)
}

main();
```

When executed as normal:
```
$ node test.js 
FIREBASE_AUTH_EMULATOR_HOST=undefined
(node:37745) UnhandledPromiseRejectionWarning: Error: Credential implementation provided to initializeApp() via the "credential" property failed to fetch a valid Google OAuth2 access token with the following error: "Error fetching access token: Error while making request: getaddrinfo ENOTFOUND metadata.google.internal metadata.google.internal:80. Error code: ENOTFOUND".
    at FirebaseAppError.FirebaseError [as constructor] (/Users/samstern/Projects/firebase-admin-node/lib/utils/error.js:43:28)
    at FirebaseAppError.PrefixedFirebaseError [as constructor] (/Users/samstern/Projects/firebase-admin-node/lib/utils/error.js:89:28)
    at new FirebaseAppError (/Users/samstern/Projects/firebase-admin-node/lib/utils/error.js:124:28)
    at /Users/samstern/Projects/firebase-admin-node/lib/firebase-app.js:122:23
    at process._tickCallback (internal/process/next_tick.js:68:7)
(node:37745) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:37745) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

When executed in the emulators:
```
$ firebase emulators:exec "node test.js"
i  emulators: Starting emulators: auth, functions
⚠  functions: The following emulators are not running, calls to these services from the Functions emulator will affect production: firestore, database, hosting, pubsub
⚠  Your requested "node" version "12" doesn't match your global version "10"
i  functions: Watching "/private/var/folders/xl/6lkrzp7j07581mw8_4dlt3b000643s/T/tmp.35bA7OYr/functions" for Cloud Functions...
i  Running script: node test.js
FIREBASE_AUTH_EMULATOR_HOST=localhost:9099
Created:  user1604932585692
✔  Script exited successfully (code 0)
i  emulators: Shutting down emulators.
i  functions: Stopping Functions Emulator
i  auth: Stopping Authentication Emulator
i  hub: Stopping emulator hub
```

### API Changes

N/A

### CODE REVIEW

File: src/auth/auth.ts [hunk 1]
```diff
@@ -822,7 +824,7 @@ export class TenantAwareAuth
 export class Auth extends BaseAuth<AuthRequestHandler>
   implements FirebaseServiceInterface, AuthInterface {
 
-  public INTERNAL: AuthInternals = new AuthInternals();
+  public INTERNAL: AuthInternals;
```
Comment (samtstern) [inline src/auth/auth.ts:827]: The reason for this change was that typescript enforces that `super()` must be the first call in the constructor if you have any inline-initialized fields.  But I have to construct the `HttpClient` before `super()`

**Review by yuchenshi — APPROVED** [review]

**Review by hiranya911 — CHANGES_REQUESTED** [review]
I'd suggest a simpler implementation like the following:

1. Change the `AuthorizedHttpClient` so child classes can override how it fetches tokens:

```typescript
export class AuthorizedHttpClient extends HttpClient {
  constructor(private readonly app: FirebaseApp) {
    super();
  }

  protected getToken(): Promise<string> {
    return this.app.INTERNAL.getToken().then((result) => {
      return result.accessToken;
    });
  }

}
```

2. Then in the Auth module extend the above class, and override `getToken()`:

```typescript
class AuthHttpClient extends AuthorizedHttpClient {
  constructor(app: FirebaseApp, private readonly useEmulator: boolean) {
    super(app);
  }

  protected getToken(): Promise<string> {
    if (this.useEmulator) {
      return Promise.resolve('owner');
    }

    return super.getToken();
  }
}

export abstract class AbstractAuthRequestHandler {
  constructor(protected readonly app: FirebaseApp, useEmulator: boolean) {
    this.httpClient = new AuthHttpClient(app, useEmulator); 
  }
}
```

**Comment by samtstern** [comment]
@hiranya911 thanks for the comments! You're right that's much cleaner.

File: src/auth/auth-api-request.ts [hunk 2]
```diff
@@ -36,6 +36,7 @@ import {
 } from './auth-config';
 import { Tenant, TenantServerResponse } from './tenant';
 import { auth } from './index';
+import { useEmulator } from './auth';
```
Comment (hiranya911) [inline src/auth/auth-api-request.ts:39]: This creates a cyclic import. Not a major problem, but something we should avoid if possible. I think one solution is to export this function from this module, and import it into the `auth` module. Another is to invoke the function in `auth`, and pass the resulting boolean value in to the classes in this module.

Also note that the constructor of `AuthResourceUrlBuilder` (in this module) also currently does the env variable check. Perhaps refactor that out with the same function while we are at it.

**Review by hiranya911 — APPROVED** [review]
LGTM with a suggestion.

**Review by hiranya911 — CHANGES_REQUESTED** [review]
Actually, can I ask you add some test cases for this change?

File: src/utils/api-request.ts [hunk 3]
```diff
@@ -814,18 +814,25 @@ export class AuthorizedHttpClient extends HttpClient {
...
       }
       return super.send(requestCopy);
     });
   }
+
+  protected getToken(): Promise<string> {
+    return this.app.INTERNAL.getToken()
+      .then((accessTokenObj) => {
+        return accessTokenObj.accessToken;
+      })
```
Comment (bojeil-google) [inline src/utils/api-request.ts:834]: Add missing semicolon.

File: src/auth/auth-api-request.ts [hunk 4]
```diff
@@ -236,6 +237,21 @@ class TenantAwareAuthResourceUrlBuilder extends AuthResourceUrlBuilder {
...
 
+/**
+ * Auth-specific HTTP client which uses the special "owner" token
+ * when communicating with the Auth Emulator.
+ */
+class AuthHttpClient extends AuthorizedHttpClient {
+
+  protected getToken(): Promise<string> {
+    if (useEmulator()) {
+      return Promise.resolve('owner');
```
Comment (bojeil-google) [inline src/auth/auth-api-request.ts:248]: Please add a unit test for this. Also why did you call it `owner`?

File: src/auth/auth-api-request.ts (hunk 2 above)
Comment (bojeil-google) [reply src/auth/auth-api-request.ts:39]: +1 to avoid cyclic import and to @hiranya911's suggestions.

File: src/utils/api-request.ts [hunk 5]
```diff
@@ -814,18 +814,25 @@ export class AuthorizedHttpClient extends HttpClient {
   }
 
   public send(request: HttpRequestConfig): Promise<HttpResponse> {
-    return this.app.INTERNAL.getToken().then((accessTokenObj) => {
+    return this.getToken().then((token) => {
       const requestCopy = Object.assign({}, request);
       requestCopy.headers = Object.assign({}, request.headers);
       const authHeader = 'Authorization';
-      requestCopy.headers[authHeader] = `Bearer ${accessTokenObj.accessToken}`;
+      requestCopy.headers[authHeader] = `Bearer ${token}`;
```
Comment (bojeil-google) [inline src/utils/api-request.ts:821]: Please add a test for the `send(request)` API when the emulator is enabled.

**Review by bojeil-google — CHANGES_REQUESTED** [review]

File: src/auth/auth-api-request.ts (hunk 4 above)
Comment (yuchenshi) [reply src/auth/auth-api-request.ts:248]: It's a convention for Firestore, RTDB, and now Auth Emulators to treat the access token `owner` as if it belonged to a service account that had IAM Role "Project Owner" (granting all permissions), hence the naming. This is documented at http://go/emulator-suite-admin-sdk.

File: src/auth/auth-api-request.ts (hunk 2 above)
Comment (samtstern) [reply src/auth/auth-api-request.ts:39]: Good idea, moved and unified.

**Review by hiranya911 — APPROVED** [review]
Thank! LGTM with a couple of nits.

**Comment by nVitius** [comment]
I know this was merged very recently, but when can we expect to see this in a tagged release?

**Comment by samtstern** [comment]
@nVitius we don't have a strict release cadence but if you look at the past we've released about once a month or so:
https://github.com/firebase/firebase-admin-node/releases

This change will be included in the next release.  If you want to use it right now you can always clone the `master` branch of this GitHub repo and `npm install` the dependency locally!

**Comment by samtstern** [comment]
@nVitius well you got lucky ... the `9.4.0` release with this fix went out!

**Pull request merged**